// react-eventsource/lib/index.js
'use strict';

exports.__esModule = true;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _reactProptypeConditionalRequire = require('react-proptype-conditional-require');

var _reactProptypeConditionalRequire2 = _interopRequireDefault(_reactProptypeConditionalRequire);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var EventSourceHOC = function (_React$Component) {
  _inherits(EventSourceHOC, _React$Component);

  function EventSourceHOC() {
    var _temp, _this, _ret;

    _classCallCheck(this, EventSourceHOC);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, _React$Component.call.apply(_React$Component, [this].concat(args))), _this), _this.state = {
      events: []
    }, _temp), _possibleConstructorReturn(_this, _ret);
  }

  EventSourceHOC.prototype.componentDidMount = function componentDidMount() {
    var _this2 = this;

    // connect to the EventSource
    this.source = this.props.source ? this.props.source : new EventSource(this.props.url);
    var cb = function cb(message) {
      _this2.setState(function (prevState) {
        var newEvents = prevState.events.concat(message.data);
        return {
          events: newEvents
        };
      });
    };
    this.props.types.forEach(function (type) {
      _this2.source.addEventListener(type, cb, false);
    });
    if (this.props.onEventSourceError) {
      this.source.onerror = this.props.onEventSourceError;
    }
  };

  EventSourceHOC.prototype.componentWillUnmount = function componentWillUnmount() {
    this.source.close();
  };

  EventSourceHOC.prototype.render = function render() {
    return _react2.default.createElement(
      'div',
      null,
      this.props.children(this.state.events)
    );
  };

  return EventSourceHOC;
}(_react2.default.Component);

process.env.NODE_ENV !== "production" ? EventSourceHOC.propTypes = {
  url: (0, _reactProptypeConditionalRequire2.default)(_propTypes2.default.string, function (props) {
    return !props.hasOwnProperty('source');
  }),
  source: (0, _reactProptypeConditionalRequire2.default)(_propTypes2.default.object, function (props) {
    return !props.hasOwnProperty('url');
  }),
  onEventSourceError: _propTypes2.default.func,
  types: _propTypes2.default.array,
  children: _propTypes2.default.func.isRequired
} : void 0;

EventSourceHOC.defaultProps = {
  types: ['message']
};

exports.default = EventSourceHOC;
module.exports = exports['default'];
